connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTask.java

```diff
@@ -396,12 +396,25 @@ boolean sendRecords() {
         for (final SourceRecord preTransformRecord : toSend) {
             ProcessingContext<SourceRecord> context = new ProcessingContext<>(preTransformRecord);
             final SourceRecord record = transformationChain.apply(context, preTransformRecord);
+            // If the result of a transformation is null, then the record should be filtered/skipped & there was no error
+            if (record == null) {
```

Comment:
The "retriable exception causes data loss" also applies to the transformations. If the transformation chain gives up retrying, it will return `null` with context.failed().

I think you can let the `null` pass through `convertTransformedRecord`, and then have separate null and context.failed() checks there that cover all of the transformation and conversion steps.

Reply:
Nice catch, updated